File: models/Gpt-3.5-turbo/py/code/mathematics/Fibonacci_Numbers.py

MOD = int(1e9) + 7

def multiply_matrices(a, b):
    # Multiply two matrices modulo MOD
    return [[(a[0][0] * b[0][0] + a[0][1] * b[1][0]) % MOD, (a[0][0] * b[0][1] + a[0][1] * b[1][1]) % MOD],
            [(a[1][0] * b[0][0] + a[1][1] * b[1][0]) % MOD, (a[1][0] * b[0][1] + a[1][1] * b[1][1]) % MOD]]

def matrix_power(matrix, n):
    # Raise a matrix to the power of n modulo MOD
    if n == 0:
        return [[1, 0], [0, 1]]  # Identity matrix
    elif n % 2 == 0:
        half_power = matrix_power(matrix, n // 2)
        return multiply_matrices(half_power, half_power)
    else:
        return multiply_matrices(matrix, matrix_power(matrix, n - 1))

def fibonacci(n):
    if n <= 1:
        return n
    matrix = [[1, 1], [1, 0]]
    result = matrix_power(matrix, n - 1)
    return result[0][0]

n = int(input())
print(fibonacci(n))

File: models/Gpt-3.5-turbo/cpp/code/problems/mathematics/Fibonacci_Numbers.cpp

#include <iostream>
#include <vector>

using namespace std;

const int MOD = 1e9 + 7;

vector<vector<long long>> matrixMultiply(const vector<vector<long long>>& A, const vector<vector<long long>>& B) {
    int n = A.size();
    int m = B[0].size();
    int k = A[0].size();

    vector<vector<long long>> C(n, vector<long long>(m, 0));

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            for (int x = 0; x < k; x++) {
                C[i][j] += (A[i][x] * B[x][j]) % MOD;
                C[i][j] %= MOD;
            }
        }
    }

    return C;
}

vector<vector<long long>> matrixPower(const vector<vector<long long>>& A, long long k) {
    if (k == 0) {
        int n = A.size();
        vector<vector<long long>> I(n, vector<long long>(n, 0));
        for (int i = 0; i < n; i++) {
            I[i][i] = 1;
        }
        return I;
    }

    vector<vector<long long>> half = matrixPower(A, k / 2);
    vector<vector<long long>> result = matrixMultiply(half, half);

    if (k % 2 == 1) {
        result = matrixMultiply(result, A);
    }

    return result;
}

long long fibonacci(long long n) {
    if (n == 0 || n == 1) {
        return n;
    }

    vector<vector<long long>> fibMatrix = {{1, 1}, {1, 0}};
    vector<vector<long long>> poweredMatrix = matrixPower(fibMatrix, n - 1);

    return poweredMatrix[0][0];
}

int main() {
    long long n;
    cin >> n;

    long long result = fibonacci(n);
    cout << result << endl;

    return 0;
}
